A42-xds-ring-hash-lb-policy.md

```diff
@@ -0,0 +1,586 @@
+the LRS policy in between those two layers as we are doing today.  So we
+need an alternative approach for handling load reporting.
+
+Initially, we considered simply building the LRS functionality directly
+into the `ring_hash` LB policy.  However, this approach has a number of
+down-sides:
+- It duplicates the LRS code in two places.  This is suboptimal but is not
+  really a huge problem; the amount of code that would need to be
+  duplicated is not large.
+- It means that LRS will work a little differently in the `ROUND_ROBIN` and
+  `RING_HASH` cases, which is confusing when we're trying to debug things.
+  It makes the system harder to understand.
+- It makes the `ring_hash` policy xDS-specific.  This is undesirable, since
+  we'd ultimately like to also use the same policy to provide affinity in
+  non-xDS use-cases.  (This will be the subject of future work in which we
+  will add fields to service config to control setting the hash for a
+  request.)
+
+So instead, we're going to completely eliminate the LRS policy and move
+all load reporting up into the `xds_cluster_impl` policy.  The approach
+will be:
+- The `xds_cluster_resolver` policy will attach an attribute to each
+  address indicating what locality the endpoint is in.
+- The `xds_cluster_impl` policy's helper will see the attribute in the
+  `CreateSubchannel()` call.  It will look up the appropriate stats
+  object and attach it to a wrapped subchannel.  If the attribute is
+  not present (e.g., because the address is from a `LOGICAL_DNS` cluster
+  instead of an EDS cluster), the stats will be reported under an empty
+  locality name.
+- When the `xds_cluster_impl` policy's picker sees the result of the pick
+  from its child, it will use the stats object in the subchannel wrapper
+  for load reporting.
+
+#### `ring_hash_experimental` LB policy
+
+We will implement a `ring_hash_experimental` LB policy that uses the same
+algorithm as Envoy's implementation.  However, this policy will be
+implemented in a non-xDS-specific way, so that it can also be used without
+xDS in the future.
+
+The `ring_hash` policy will accept both multiple instances of the same
+address and the per-address weight attribute.  This provides two
+different ways to indicate weighting when generating an address list.
+
+Whenever the address list changes, a new ring will be generated.  The
+ring will be stored within the picker, so any time a new ring is
+generated, a new picker will be returned.
+
+The picker will contain not just the ring but also the current state of
+every subchannel in the ring.  Whenever a subchannel's state changes, a
+new picker will need to be generated.
+
+##### LB Policy Config
+
+The `ring_hash_experimental` policy will have the following config:
+
+```
+message RingHashLoadBalancingConfig {
+  uint64 min_ring_size = 1;
+  uint64 max_ring_size = 2;
+}
+```
+
+These parameters are used to determine how the ring is constructed, just
```

Comment:
May worth mentioning using defaults (1024 for min_ring_size and 8M for max_ring_size) if the config/values do not present.

Reply:
Maybe this should be applied at the time when receiving the values in CDS response, as those are "xDS default". Just from ring_hash LB policy's perspective (not xDS-specific), it's gRPC's own choice whether to use default and what default values are. I think same rule should apply for circuit_breaker's default value as well.

Reply:
I've clarified that the LB policy config uses the same defaults as xDS.

I've also noted that the `ring_hash_lb_config` field is optional in the CDS resource.